src/measure_theory/decomposition/signed_hahn.lean

```diff
@@ -0,0 +1,676 @@
+  restricted on `i` is non-positive, i.e. `v.restrict i ≤ 0`.
+
+## Main results
+
+* `signed_measure.exists_disjoint_positive_negative_union_eq` : the Hahn decomposition theorem.
+* `signed_measure.exists_negative_set` : A measurable set of negative measure contains at least
+  one negative subset.
+
+## Tags
+
+positive, negative, Hahn decomposition theorem
+-/
+
+noncomputable theory
+open_locale classical big_operators nnreal ennreal
+
+section lemmas
+
+variables {α : Type*}
+
+open_locale topological_space
+
+open set filter
+
+lemma exists_tendsto_Inf {S : set ℝ} (hS : S.nonempty) (hS' : bdd_below S) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (𝓝 (Inf S)) :=
+begin
+  have : ∀ n : ℕ, ∃ t ∈ S, t < Inf S + 1 / (n + 1 : ℝ),
+  { exact λ n, (real.Inf_lt _ hS hS').1 ((lt_add_iff_pos_right _).2 nat.one_div_pos_of_nat) },
+  choose f hf using this,
+  refine ⟨f, λ n, (hf n).1, _⟩,
+  rw tendsto_iff_dist_tendsto_zero,
+  refine squeeze_zero_norm _ tendsto_one_div_add_at_top_nhds_0_nat,
+  intro n,
+  obtain ⟨hf₁, hf₂⟩ := hf n,
+  rw [real.dist_eq, real.norm_eq_abs, abs_abs,
+      abs_of_nonneg (sub_nonneg.2 (real.Inf_le S hS' hf₁))],
+  linarith,
+end
+
+lemma exists_tendsto_Sup {S : set ℝ} (hS : S.nonempty) (hS' : bdd_above S) :
+  ∃ (f : ℕ → ℝ) (hf : ∀ n, f n ∈ S), tendsto f at_top (𝓝 (Sup S)) :=
+begin
+  have : ∀ n : ℕ, ∃ t ∈ S, Sup S - 1 / (n + 1 : ℝ) < t,
+  { intro n,
+    apply (real.lt_Sup _ hS hS').1,
+    rw [sub_lt, _root_.sub_self],
+    exact nat.one_div_pos_of_nat },
+  choose f hf using this,
+  refine ⟨f, λ n, (hf n).1, _⟩,
+  rw tendsto_iff_dist_tendsto_zero,
+  refine squeeze_zero_norm _ tendsto_one_div_add_at_top_nhds_0_nat,
+  intro n,
+  obtain ⟨hf₁, hf₂⟩ := hf n,
+  rw [real.dist_eq, real.norm_eq_abs, abs_abs, abs_of_nonpos],
+  { linarith },
+  { rw [sub_le, _root_.sub_zero],
+    exact real.le_Sup S hS' hf₁ },
+end
+
+end lemmas
+
+variables {α β : Type*} [measurable_space α]
+variables {M : Type*} [add_comm_monoid M] [topological_space M] [ordered_add_comm_monoid M]
+
+namespace measure_theory
+
+namespace vector_measure
+
+/-- A set `i` is positive with respect to a vector measure `v` if `v` restricted
+on`i` is non-negative, i.e. `0 ≤ v.restrict i`. -/
+def positive (v : vector_measure α M) (i : set α) : Prop :=
+0 ≤ v.restrict i
+
+lemma positive_iff {v : vector_measure α M} {i : set α} (hi : measurable_set i) :
+  v.positive i ↔ ∀ ⦃j⦄, measurable_set j → j ⊆ i → 0 ≤ v j :=
+begin
+  convert restrict_le_restrict_iff 0 v hi,
+  rw restrict_zero, refl,
+end
+
+/-- A set `i` is negative with respect to a vector measure `v` if `v` restricted
+on`i` is non-positive, i.e. `v.restrict i ≤ 0`. -/
+def negative (v : vector_measure α M) (i : set α) : Prop :=
+v.restrict i ≤ 0
+
+lemma negative_iff {v : vector_measure α M} {i : set α} (hi : measurable_set i) :
+  v.negative i ↔ ∀ ⦃j⦄, measurable_set j → j ⊆ i → v j ≤ 0 :=
+begin
+  convert restrict_le_restrict_iff v 0 hi,
+  rw restrict_zero, refl,
```

Comment:
It's annoying that `0.restrict i` is not defeq to `0`, but I guess there's not much we can do about that...